### ISSUE TITLE
Is there a clean way to access the conan database from a recipe?

### ISSUE THREAD

**Issue by Stefan-working** [conan-io/conan#5827]
To help us debug your issue please explain:

- [  x ] I've read the [CONTRIBUTING guide](https://github.com/conan-io/conan/blob/develop/.github/CONTRIBUTING.md).
- [  x ] I've specified the Conan version, operating system version and any tool that can be relevant.
- [ x ] I've explained the steps to reproduce the error or the motivation/use case of the question/suggestion.

I have the issue that I want to access data from a generic repository on the same artifactory as the conan repository. Therefore currently I have added this dirty workaround:
```python
        def getConanDatabase(self):
            color = colorama_initialize()
            out = ConanOutput(sys.stdout, sys.stderr, color)
            user_home = get_conan_user_home()
            base_folder = os.path.join(user_home, ".conan")

            cache = ClientCache(base_folder, out)
            localdb = LocalDB.create(cache.localdb)
            return localdb

# some other code

        user, token = self._conanDB.get_login(self.conanURL) 

        head = {'Authorization': "Bearer %s" % str(token)}
        try:
            tools.get(fullUrl, destination=self._test_data_folder,
                      headers=head)
        except ConanException as ex:
            print("Raw Data Loading Error: ", ex)
```
While this works good in the current situation I assume that there is a high probability that this will break in the future due to conan updates and the fact that I just hijacked the necessary conan functions to access this database. Is there currently a better way to achive the same thing?

**Comment by uilianries** [comment]
Hi @Stefan-working !

What's the propose of this?

Conan offers an [API](https://github.com/conan-io/conan/blob/develop/conans/client/conan_api.py) which contains many useful methods, including remote requests. The API is not documented because it's not on stable version, which means, it can be changed from version to version, but it's safe for regular usage, since Conan client is using it.

About the data base, I don't recommend doing it, it's invasive and probably you need to update your code for any new change on Conan side.

**Comment by Stefan-working** [comment]
Hello  uilianries,
thank you for your fast response. 
We have huge image and video data which is not part of a conan repository but of a generic repository within artifactory. This data is used for testing.
Therefore maybe you can give me a hint how I could use the remote api as it is only designed to be used for a conan remote if I am correct?

What I actually do at the moment is using the token for the artifactory, which is the same for the conan repository, a generic repository or any other repository within the same artifactory. This way it is not necessary to authenticate again at the artifactory. 
Of course if there would be a way for the get function to just use known authentification for a already known base url this would be the most convenient solution.

**Comment by uilianries** [comment]
If your generic package is only for testing, why do you need a login for it? Maybe you could expose it in public repo in Artifactory, so you don't need the token for this package.

Usually my token as store in files which only my user has access. Also, I have a bunch of scripts which read my access token. Your case is special, you are sharing the same token. What about an env var?

**Comment by memsharded** [comment]
> Of course if there would be a way for the get function to just use known authentication for a already known base url this would be the most convenient solution.

The problem I see with this solution, is that this is not necessarily always true. Different repos can have different access policies, permissions, etc., so the token available for Conan might not work for a generic repo or viceversa. We could have a deeper look, but if the solution is not general enough, then implementing it as a built-in, is not possible. Same applies for reading the token directly from the local DB, that token still might not be valid for a generic repo, so enabling reading the DB token still feels a bit hacky, even if the access was simplified.

I think that if I had to do it, I'd try to be pragmatic and keep it as simple as possible, have just some read-only credentials for that test generic repo, and pass them in env-var. 

Another alternative would be to create conan packages with those files inside, that would have some beneficial advantages, for example if you are using the same data as test input for different packages, that would be cached in the conan cache, saving tons of download time. Have you considered this approach?

**Comment by Stefan-working** [comment]
Thank you for your responses 

So am I correct that a clean solution would be to enable the usual conan authentification routines (asking for username, password and so on) for none conan repositories and enable get to ask for credentials? This could be also useful for git source repositories or similar stuff. But I assume as such a solution may take time to implement that it won't come soon if at all?

Yeah a binary only package is the other thing I consider using. So a recipe for this approach should be quiet similar to a recipe for a header only library? (https://docs.conan.io/en/latest/howtos/header_only.html) 
To use the test data in another project I would just require them as a build_requirement so that post build tests can be executed.

Using an env var may be also possible but actual I don't think it is easy enough. The idea is that a new developer could just check out a current version and everything should -thank you to conan- just work without any additional necessary steps;)

To sum it up there is no clean way to access an access token and most likely there won't be one in the near future.

**Comment by uilianries** [comment]
> Yeah a binary only package is the other thing I consider using. So a recipe for this approach should be quiet similar to a recipe for a header only library? (docs.conan.io/en/latest/howtos/header_only.html)

In fact, but there is a better approach:
https://docs.conan.io/en/latest/creating_packages/existing_binaries.html

> Using an env var may be also possible but actual I don't think it is easy enough. The idea is that a new developer could just check out a current version and everything should -thank you to conan- just work without any additional necessary steps;)

You can use profiles with env vars embedded, so you are able to share any profile:

https://docs.conan.io/en/latest/mastering/sharing_settings_and_config.html
https://docs.conan.io/en/latest/reference/profiles.html#package-settings-and-env-vars

**Comment by Stefan-working** [comment]
Hello uilianries,
can you please explain why the binary recipe variant is better than the header only variant?
As I see it images and videos are invariant for all platforms, which is also the case vor header only libraries but not for the binary packages as these may contain build tools or other platform dependend stuff.

**Comment by uilianries** [comment]
Hi @Stefan-working !

Please, take a look on docs.conan.io/en/latest/creating_packages/existing_binaries.html

To summarize, when are packaging pre-built binaries, maybe some settings are important, like os and arch. Also, you don't need to use header_only when packaging pre-built binaries.

**Comment by Stefan-working** [comment]
Okay thank you, this just confirms, what I already wrote - as in my case everything is about  images and videos which are os and arch independent the header only variant should be the cleaner one;) No additional copy has to be done every platform will have the exactly same data and no additional packages are created / necessary which is usually not the case for the binary variant.

**Comment by memsharded** [comment]
> To sum it up there is no clean way to access an access token and most likely there won't be one in the near future.

Yes, I am afraid that is the case, seems a complicated thing with too many potential corner cases.

If you are going to try wrapping them in packages that can be used as build-requires, that sounds good. Let us know how it goes and if you have any further question. Thanks!
